core/kazoo_number_manager/src/carriers/knm_voxbone.erl

```diff
@@ -0,0 +1,332 @@
+        {'error', Message} ->
+            _ = maybe_destroy_cart(Cart),
+            knm_errors:unspecified(Message, 503)
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc checkout the order and generate the DID summary
+%% @end
+%%------------------------------------------------------------------------------
+-spec checkout(map()) -> {'error', any()} | {'ok', knm_number:phone_numbers()}.
+checkout(#{id := CartId, query_id := _QueryId}=Cart) ->
+    case knm_voxbone_util:voxbone_request('get', <<"ordering/cart/",(kz_term:to_binary(CartId))/binary,"/checkout">>, []) of
+        {'error', Message} ->
+            _ = maybe_destroy_cart(Cart),
+            knm_errors:unspecified(Message, 503);
+        {'ok',  Response} ->
+            generate_number_summary(Cart, kz_json:get_value(<<"productCheckoutList">>, Response))
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc extract actual DIDs from order and normalize to knm search results
+%% @end
+%%------------------------------------------------------------------------------
+-spec generate_number_summary(map(), kz_json:object()) -> {'error', any()} | {'ok', knm_number:numbers()}.
+generate_number_summary(#{id := _CartId, query_id := _QueryId}=Cart, OrderSummary) ->
+    generate_number_summary(Cart, OrderSummary, []).
+
+-spec generate_number_summary(map(), kz_json:object(), kz_term:objects()) -> {'error', any()} | {'ok', knm_number:numbers()}.
+generate_number_summary(#{id := _CartId, query_id := _QueryId}=_Cart, [], Acc) ->
+    {'ok', Acc};
+generate_number_summary(#{id := _CartId, query_id := QueryId}=Cart, [Order | Rest], Acc) ->
+    case kz_json:get_value(<<"productType">>, Order) =:= <<"DID">> of
+        'false' ->
+            generate_number_summary(Cart, Rest, Acc);
+        'true' ->
+            Options = [{<<"orderReference">>, kz_json:get_value(<<"orderReference">>, Order)}
+                      ] ++ knm_voxbone_util:required_params(),
+            {'ok', DIDs} = knm_voxbone_util:voxbone_request('get', <<"inventory/did">>, Options),
+            KNMs = order_to_KNM(QueryId, DIDs),
+            generate_number_summary(Cart, Rest, KNMs ++ Acc)
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc convert voxbone dids into knm phone numbers
+%% @end
+%%------------------------------------------------------------------------------
+-spec order_to_KNM(kz_term:ne_binary(), kz_json:objects()) -> [{knm_search:option(), {kz_term:ne_binary(), kz_term:atom(), kz_term:atom(), kz_json:object()}}].
+order_to_KNM(QueryId, DIDs) ->
+    JObjs = kz_json:get_value(<<"dids">>, DIDs),
+    order_to_KNM(QueryId, JObjs, []).
+
+-spec order_to_KNM(kz_term:ne_binary(), kz_json:objects(), knm_number:phone_numbers()) -> [{knm_search:option(), {kz_term:ne_binary(), kz_term:atom(), kz_term:atom(), kz_json:object()}}].
```

Comment:
Is `knm_number:phone_numbers()` a real type? Or is this `kz_term:ne_binary()` for the DID?